### PR TITLE
Add "Render as a surface" option to 2D point cloud renderers

### PR DESCRIPTION
**Pull request by wonder-sk**
When enabled, we will do Delaunay triangulation of the points in the current map view, and then render triangles instead of points. For each point we keep its color for interpolation in the triangle.

Global map shading is also supported with the new option, when enabled, we also keep elevation of each point, and then rasterize triangles with interpolated elevations to the provided elevation map.

When "Render as a surface" is enabled, drawing order is ignored, because points do not obscure other points anymore - all input points participate in the triangulation.

There is also an option to filter large triangles (given by the maximum length of edge of a triangle), which is useful when one wants to see the actual holes in the data. Compared to the implementation for 3D rendering, the 2D rendering only provides filtering based on horizontal length of triangles. Filtering based on triangle size on the vertical axis seems irrelevant because the 2D view is always from the top.

Funded by: National Resources Canada

## GUI

There's a new group in 2D styling tab - similar to the one that already exists in the 3D styling tab:

pc-2d-surface-gui

## Screenshots

As expected, triangulation covers holes between points when zoomed in, but subjectively also gives nicer rendering when zoomed out.

**Trencin castle**

| Disabled        | Enabled |
| ------------- | ------------- |
| pc-2d-surface-trencin-full-1      | pc-2d-surface-trencin-full-2 |
| pc-2d-surface-trencin-zoom-1      | pc-2d-surface-trencin-zoom-2 |

Global map shading when render as a surface is enabled:

| With map shading | With map shading, only ground, filtering large triangles  |
| ------------- | ------------- |
| pc-2d-surface-trencin-full-2-shading | pc-2d-surface-trencin-full-2-ground-shading-filter |

**Helsinki train station (RGB)**

| Disabled        | Enabled |
| ------------- | ------------- |
| pc-2d-surface-helsinki-1 | pc-2d-surface-helsinki-2 |
| pc-2d-surface-helsinki-full-1 | pc-2d-surface-helsinki-full-2 |

### CODE REVIEW

File: src/core/pointcloud/qgspointcloudlayerrenderer.cpp [hunk 1]
```diff
@@ -569,6 +592,129 @@ int QgsPointCloudLayerRenderer::renderNodesSorted( const QVector<IndexedPointClo
...
+  QSize outputSize = img.size();
+
+  int topLim = std::max( int( screenBBox.yMinimum() ), 0 );
+  int bottomLim = std::min( int( screenBBox.yMaximum() ), outputSize.height() - 1 );
+  int leftLim = std::max( int( screenBBox.xMinimum() ), 0 );
+  int rightLim = std::min( int( screenBBox.xMaximum() ), outputSize.width() - 1 );
+
+  int red0 = qRed( c0 ), green0 = qGreen( c0 ), blue0 = qBlue( c0 );
+  int red1 = qRed( c1 ), green1 = qGreen( c1 ), blue1 = qBlue( c1 );
+  int red2 = qRed( c2 ), green2 = qGreen( c2 ), blue2 = qBlue( c2 );
```
Comment (nyalldawson) [inline src/core/pointcloud/qgspointcloudlayerrenderer.cpp:623]: There's an opportunity here for an optimisation -- we could check if c0 == c1 == c2 (and not elevation map), and if so then just use the QPainter renderPolygon method to draw a filled solid color triangle. 

It's a situation that WILL arise frequently, eg for classified point clouds.

File: src/core/pointcloud/qgspointcloudrenderer.h [hunk 2]
```diff
@@ -673,6 +798,11 @@ class CORE_EXPORT QgsPointCloudRenderer
     Qgis::PointCloudSymbol mPointSymbol = Qgis::PointCloudSymbol::Square;
     int mPainterPenWidth = 1;
     Qgis::PointCloudDrawOrder mDrawOrder2d = Qgis::PointCloudDrawOrder::Default;
+
+    bool mRenderAsTriangles = false;
+    bool mHorizontalTriangleFilter = false;
+    float mHorizontalTriangleFilterThreshold = 5.0;
```
Comment (nyalldawson) [inline src/core/pointcloud/qgspointcloudrenderer.h:804]: I think this should be double -- the internal calculations can use float, but it's weird for user-exposed values to eg change from 1.20 to 1.1999999 in the GUI.

File: src/core/pointcloud/qgspointcloudlayerrenderer.cpp (hunk 1 above)
Comment (wonder-sk) [reply src/core/pointcloud/qgspointcloudlayerrenderer.cpp:623]: I have tried this, but unfortunately it leaves ugly artifacts at the edges between barycentric-based and QPainter-based triangles - even when I use 1px solid pen in addition to the brush fill for QPainter, and I would prefer not to use thicker pens. After this initial work, I would like to do a bit of profiling, so if possible I would prefer to leave optimisation for later (moreover, for attribute by ramp or RGB, the QPainter solid triangle path would not get triggered often)

image

File: src/core/pointcloud/qgspointcloudrenderer.h (hunk 2 above)
Comment (wonder-sk) [reply src/core/pointcloud/qgspointcloudrenderer.h]: Done

File: src/core/pointcloud/qgspointcloudlayerrenderer.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/pointcloud/qgspointcloudlayerrenderer.cpp:623]: Ok, fair enough!

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**